### PR TITLE
Setup mechanism for register customized query DSL logic

### PR DESCRIPTION
**Pull request by SuZhou-Joe**
### Check List

- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
  - [ ] `yarn test:ftr`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] Update [CHANGELOG.md](./../CHANGELOG.md)
- [x] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: src/core/server/saved_objects/service/lib/repository.ts [hunk 1]
```diff
@@ -1766,6 +1769,122 @@ export class SavedObjectsRepository {
     };
   }
 
+  async getPermissionQueryDSL(props: {
```
Comment (wanglam) [inline src/core/server/saved_objects/service/lib/repository.ts:1772]: Will these two new repository methods be used in other places? For now, they are only be called in the saved object client. Is that means we will call  `getPermissionQueryDSL` or  `processFindOptions` directly saved object client in the real case?

Comment (xluo-aws) [reply src/core/server/saved_objects/service/lib/repository.ts:1772]: Should we name it QueryDSL? If the data will not be stored in open search?

Comment (SuZhou-Joe) [reply src/core/server/saved_objects/service/lib/repository.ts:1772]: > Will these two new repository methods be used in other places? For now, they are only be called in the saved object client. Is that means we will call `getPermissionQueryDSL` or `processFindOptions` directly saved object client in the real case?

There is no difference to call these two methods through savedObjectsClient or repositoryClient, do you want to say that we should define these two methods only in repositoryClient?

Comment (SuZhou-Joe) [reply src/core/server/saved_objects/service/lib/repository.ts:1772]: > Should we name it QueryDSL? If the data will not be stored in open search?

I think it is OK to name it `QueryDSL` as it stands for "query domain specific language" here and it is not a opensearch specific term.

File: src/core/server/saved_objects/service/lib/repository.ts [hunk 2]
```diff
@@ -1766,6 +1769,122 @@ export class SavedObjectsRepository {
...
+        permissionTypes: options.permissionModes ?? [
+          WorkspacePermissionMode.LibraryRead,
+          WorkspacePermissionMode.LibraryWrite,
+          WorkspacePermissionMode.Management,
+        ],
+        principals,
+        savedObjectType: [WORKSPACE_TYPE],
+      });
+    } else {
+      const permittedWorkspaceQueryDSL = await this.getPermissionQueryDSL({
```
[the remark this answers is not in the document]
Comment (SuZhou-Joe) [reply src/core/server/saved_objects/service/lib/repository.ts:1801]: Do you mean that we should exact getPermittedWorkspaceIds to a private methods of repository?

[the remark this answers is not in the document]
Comment (wanglam) [reply src/core/server/saved_objects/service/lib/repository.ts:1801]: It's a nice to have. I am prefer to separate an util function, and take one repository object as parameter. Then we can reuse the same logic.

[the remark this answers is not in the document]
Comment (SuZhou-Joe) [reply src/core/server/saved_objects/service/lib/repository.ts:1801]: That works, will do that.

**Review by wanglam — APPROVED** [review]

**Pull request merged**